modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java

```diff
@@ -0,0 +1,1986 @@
+                        SnapshotResponseMessage respMsg0 = (SnapshotResponseMessage)msg;
+
+                        RemoteSnapshotFuture fut0 = rmtSnpReq.get();
+
+                        if (fut0 == null || !fut0.snpName.equals(respMsg0.snapshotName())) {
+                            if (log.isInfoEnabled()) {
+                                log.info("A stale snapshot response message has been received. Will be ignored " +
+                                    "[fromNodeId=" + nodeId + ", response=" + respMsg0 + ']');
+                            }
+
+                            return;
+                        }
+
+                        if (respMsg0.errorMessage() != null) {
+                            fut0.onDone(new IgniteCheckedException("Request cancelled. The snapshot operation stopped " +
+                                "on the remote node with an error: " + respMsg0.errorMessage()));
+                        }
+                    }
+                }
+                catch (Throwable e) {
+                    U.error(log, "Processing snapshot request from remote node fails with an error", e);
+
+                    cctx.kernalContext().failure().process(new FailureContext(FailureType.CRITICAL_ERROR, e));
+                }
+                finally {
+                    busyLock.leaveBusy();
+                }
+            }
+        });
+
+        cctx.gridEvents().addDiscoveryEventListener(discoLsnr = (evt, discoCache) -> {
+            if (!busyLock.enterBusy())
+                return;
+
+            try {
+                UUID leftNodeId = evt.eventNode().id();
+
+                if (evt.type() == EVT_DISCOVERY_CUSTOM_EVT) {
+                    DiscoveryCustomEvent evt0 = (DiscoveryCustomEvent)evt;
+
+                    if (evt0.customMessage() instanceof InitMessage) {
+                        InitMessage<?> msg = (InitMessage<?>)evt0.customMessage();
+
+                        // This happens when #takeSnapshot() method already invoked and distributed process
+                        // starts its action.
+                        if (msg.type() == START_SNAPSHOT.ordinal()) {
+                            assert clusterSnpRq != null ||
+                                !CU.baselineNode(cctx.localNode(), cctx.kernalContext().state().clusterState()) : evt;
+
+                            DiscoveryCustomEvent customEvt = new DiscoveryCustomEvent();
+
+                            customEvt.node(evt0.node());
+                            customEvt.eventNode(evt0.eventNode());
+                            customEvt.affinityTopologyVersion(evt0.affinityTopologyVersion());
+                            customEvt.customMessage(new SnapshotStartDiscoveryMessage(discoCache, msg.processId()));
+
+                            // Handle new event inside discovery thread, so no guarantees will be violated.
+                            cctx.exchange().onDiscoveryEvent(customEvt, discoCache);
```

Comment:
I don't like the idea of injecting to the discovery and using an internal `onDiscoveryEvent` method. I think we can do it with only one event. For example,  `SnapshotStartDiscoveryMessage` can extend `InitMessage` and we can pass `InitMessage` class and `InitMessage` factory to `DistributedProcess` constructor. In this case, you don't need to make `onDiscoveryEvent` public and can use standard discovery workflow.

Reply:
Discussed privately. A factory added to the `DistributedProcess` configuration, so it is possible from now start a process with custom inheritor of `InitMessage`, including `SnapshotDiscoveryMessage`.